api/src/main/java/org/apache/iceberg/util/StructProjection.java

```diff
@@ -134,11 +156,17 @@ public int size() {
 
   @Override
   public <T> T get(int pos, Class<T> javaClass) {
+    int structPos = positionMap[pos];
+
     if (nestedProjections[pos] != null) {
-      return javaClass.cast(nestedProjections[pos].wrap(struct.get(positionMap[pos], StructLike.class)));
+      return javaClass.cast(nestedProjections[pos].wrap(struct.get(structPos, StructLike.class)));
     }
 
-    return struct.get(positionMap[pos], javaClass);
+    if (structPos != -1) {
+      return struct.get(structPos, javaClass);
+    } else {
+      return null;
```

Comment:
Since this relies on a `null` for `nestedProjections[pos]`, I think it should set `nestedProjections[pos]` to `null` in the constructor where `positionMap[pos]` is set to -1.

Reply:
I think it is actually always null as long as the field was not found. I'll add an explicit call, though.

Reply:
Added.